sound/soc/sof/core.c

```diff
@@ -353,24 +339,28 @@ int snd_sof_device_remove(struct device *dev)
 	if (IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE))
 		cancel_work_sync(&sdev->probe_work);
 
-	snd_sof_fw_unload(sdev);
-	snd_sof_ipc_free(sdev);
-	snd_sof_free_debug(sdev);
-	snd_sof_free_trace(sdev);
+	if (sdev->fw_state > SOF_FW_BOOT_NOT_STARTED) {
+		snd_sof_fw_unload(sdev);
+		snd_sof_ipc_free(sdev);
+		snd_sof_free_debug(sdev);
+		snd_sof_free_trace(sdev);
+	}
```

Comment:
that looks incorrect to me. In case of errors during run_firmware, you will jump to fw_run_err and then you only mark things SOF_FW_BOOT_NOT_STARTED, so here you would never call snd_sof_fw_unload.

In other words, we need to keep track of what resources were successfully allocated before rolling back and freeing them. This test case does not work I am afraid.

that said, I think your solution for suspend is good, I don't think it's good enough to only release resources in probe_continue(). we could still see a suspend and fail because of unallocated resources.

Reply:
@plbossart I think it's the reverse. In probe, we either reserve all resource or we reserve none. Thus we only need to track whether all resources are allocated or not, and if yes, free them here.
In case run_firmware fails, the goto labels in sof_probe_continue() ensure only those resources are freed which were allocated. At the end, if it failed and state goes back to NOT_STARTED; then we have freed all resources. Right?

Reply:
Let's take an example

````
fw_trace_err:
	snd_sof_free_trace(sdev);
fw_run_err: <<<< you land here on a probe error
	snd_sof_fw_unload(sdev);
fw_load_err:
	snd_sof_ipc_free(sdev);
ipc_err:
	snd_sof_free_debug(sdev);
dbg_err:
	snd_sof_remove(sdev);
````

and do this in remove


````
if (sdev->fw_state > SOF_FW_BOOT_NOT_STARTED) {
		snd_sof_fw_unload(sdev);
		snd_sof_ipc_free(sdev);
		snd_sof_free_debug(sdev);
<<< all the 3 above were already released
		snd_sof_free_trace(sdev);
<<< only this one is necessary.
	}
````

So I can't follow the all or none explanation, it's not what the code says?

Reply:
@plbossart wrote:
> Let's take an example
```
fw_run_err: <<<< you land here on a probe error
	snd_sof_fw_unload(sdev);
[...]
 dbg_err:
 	snd_sof_remove(sdev);
// KV: and the patch also has:
        fw_state = SOF_FW_BOOT_NOT_STARTED
// KV: traces were not allocated, and for rest of the three we unload, ipc_free, free_debug and sof_remove, so all resoruces are freed no matter which point the probe failed 
```

> and do this in remove
```
 if (sdev->fw_state > SOF_FW_BOOT_NOT_STARTED) {
// KV: we don't run this branch at all as state is BOOT_NOT_STARTED as set in the probe error handler
 		snd_sof_fw_unload(sdev);
 		snd_sof_ipc_free(sdev);
 		snd_sof_free_debug(sdev);
 <<< all the 3 above were already released
 		snd_sof_free_trace(sdev);
 <<< only this one is necessary.
// KV: as the probe failed to run, we never allocated trace
 	}
 ```

Makes sense?

Reply:
ok, got it now, thanks @kv2019i for bearing with me. Indeed, because we've already rolled back all allocations in the error handling there's nothing to do in the .remove.
Nice indeed, states would only be needed if the errors were handled in a non-sequential manner. let's close #1743 and #1745 then

Fine with me, I'll let you battle with @ranj063 for the other details.